irods/auth/__init__.py

```diff
@@ -1,8 +1,117 @@
+import importlib
+import logging
+import weakref
+from irods.api_number import api_number
+from irods.message import iRODSMessage, JSON_Message
+import irods.password_obfuscation as obf
+import irods.session
+
+
 __all__ = ["pam_password", "native"]
 
+
 AUTH_PLUGIN_PACKAGE = "irods.auth"
 
-import importlib
+
+# Python3 does not have types.NoneType
+_NoneType = type(None)
+
+
+class AuthStorage:
+    """A class that facilitates flexible means password storage.
+
+    Using an instance of this class, passwords may either be
+
+        - directly placed in a member attribute (pw), or
+
+        - they may be written to / read from a specified file path in encoded
+          form, usually in an .irodsA file intended for iRODS client authentication.
+
+    Most typical of this class's utility is the transfer of password information from
+    the pam_password to the native authentication flow.  In this usage, whether the
+    password is stored in RAM or in the filesystem depends on whether it was read
+    originally as a function parameter or from an authentication file, respectively,
+    when the session was created.
+
+    """
+
+    @staticmethod
+    def get_env_password(filename = None):
+        options = dict(irods_authentication_file = filename) if filename else {}
+        return irods.session.iRODSSession.get_irods_password(**options)
+
+    @staticmethod
+    def get_env_password_file():
+        return irods.session.iRODSSession.get_irods_password_file()
+
+    @staticmethod
+    def set_env_password(unencoded_pw, filename = None):
+        if filename is None:
+            filename = AuthStorage.get_env_password_file()
+        from ..client_init import _open_file_for_protected_contents
+        with _open_file_for_protected_contents(filename,'w') as irodsA:
+            irodsA.write(obf.encode(unencoded_pw))
+        return filename
+
+    @staticmethod
+    def get_temp_pw_storage(conn):
+        """Fetch the AuthStorage instance associated with this connection object.
+        """
+        return getattr(conn,'auth_storage',lambda:None)()
+
+    @staticmethod
+    def create_temp_pw_storage(conn):
+        """Creates an AuthStorage instance to be cached and associated with this connection object.
+
+           Called multiple times for the same connection, it will return the cached instance.
+
+           The value returned by this call should be stored by the caller into an appropriately scoped
+           variable to ensure the AuthStorage instance endures for the desired lifetime -- that is,
+           for however long we wish to keep the password information around.  This is because the
+           connection object only maintains a weak reference to said instance.
+        """
+
+        # resolve the weakly referenced AuthStore obj for the connection if there is one.
+        weakref_to_store = getattr(conn,'auth_storage',None)
+        store = (weakref_to_store and weakref_to_store())
```

Comment:
Doesn't this result in a boolean - i.e. it can never be `None`?

Reply:
That does not result in a boolean:
```python
>>> def foo():
...     return 42
... 
>>> x = foo
>>> y = None
>>> x and x()
42
>>> y and y()
>>> z = x and x()
>>> z
42
>>> z = y and y()
>>> z
>>> z is None
True
```

Reply:
That breaks my brain, but good.

Resolving.